### PR TITLE
Checkbox: Support inferring of tri-state checked value

### PR DESCRIPTION
**Pull request by michaeltaranto**
**Checkbox:** Support inferring of tri-state checked value

To simplify the use of tri-state checkboxes, the **checked** prop now supports resolving the tri-state value from a array of checked values.

**EXAMPLE USAGE:**
```jsx
 <Checkbox
  label="Select all"
  checked={[ true, false, false ]} // Will resolve to "mixed"
/>
```

### CODE REVIEW

File: lib/components/Checkbox/Checkbox.test.tsx [hunk 1]
```diff
@@ -195,3 +196,25 @@ describe('Checkbox', () => {
     expect(checkbox.checked).toBe(false);
   });
 });
+
+describe('resolveCheckedGroup', () => {
```
Comment (markdalgleish) [inline lib/components/Checkbox/Checkbox.test.tsx:200]: Feel like we should be testing via the public API rather than an internal util. What do you think?

Comment (michaeltaranto) [reply lib/components/Checkbox/Checkbox.test.tsx:200]: Yeah i think you're right. Legacy of previous iterations 😄

File: .changeset/nervous-papayas-warn.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  - Checkbox
+---
+
+**Checkbox:** Support inferring of tri-state checked value
+
+To simplify the use of tri-state checkboxes, the **checked** prop now supports resolving the tri-state value from a array of checked values.
+
+**EXAMPLE USAGE:**
+```jsx
+// Will resolve to "mixed" based on the array of values provided
```
Comment (markdalgleish) [inline .changeset/nervous-papayas-warn.md:16]: Nit: Move this comment down, on the same line as the checked prop?

File: lib/components/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -5,12 +5,42 @@ import {
...
+    if (acc === true) {
+      return val || 'mixed';
+    }
+
+    if (acc === false && index > 0) {
+      return val ? 'mixed' : false;
+    }
+
+    return val;
+  }, false);
```
Comment (BurntCaramel) [inline lib/components/Checkbox/Checkbox.tsx:23]: Sorry to jump in, but it did make me think that this could be shortened to something like:

```suggestion
  values.some(val => val !== values[0]) ? 'mixed' : a[0] ?? false;
```

Comment (michaeltaranto) [reply lib/components/Checkbox/Checkbox.tsx:23]: Sometimes you just stare at the same code for too long 😄 .

Thanks mate, i like the refactor 👏

File: .changeset/nervous-papayas-warn.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+---
+
+---
+updated:
+  - Checkbox
+---
+
+**Checkbox:** Support inferring of tri-state checked value
+
+To simplify the use of tri-state checkboxes, the **checked** prop now supports resolving the tri-state value from a array of checked values.
```
Comment (markdalgleish) [inline .changeset/nervous-papayas-warn.md:12]:
```suggestion
To simplify the use of tri-state checkboxes, the **checked** prop now supports resolving the tri-state value from an array of checked values.
```

**Comment by markdalgleish** [comment]
Might be worth adding a test for an empty array, just so we can't accidentally change the behaviour in the future?

**Review by markdalgleish — APPROVED** [review]
Just one minor typo fix and a question about adding another test, but happy to leave it with you from here. This is awesome 🙌
